clippy_lints/src/bool_to_int_with_if.rs

```diff
@@ -0,0 +1,123 @@
+    /// ### Why is this bad?
+    /// Coercion or `from()` is idiomatic way to convert bool to a number.
+    /// Both methods are guaranteed to return 1 for true, and 0 for false. See https://doc.rust-lang.org/std/primitive.bool.html#impl-From%3Cbool%3E
+    ///
+    /// ### Example
+    /// ```rust
+    /// let condition = false;
+    /// if condition {
+    ///     1_i64
+    /// } else {
+    ///     0
+    /// };
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// let condition = false;
+    /// i64::from(condition);
+    /// ```
+    /// or
+    /// ```rust
+    /// let condition = false;
+    /// condition as i64;
+    /// ```
+    #[clippy::version = "1.64.0"]
+    pub BOOL_TO_INT_WITH_IF,
+    style,
+    "using if to convert bool to int"
+}
+declare_lint_pass!(BoolToIntWithIf => [BOOL_TO_INT_WITH_IF]);
+
+impl<'tcx> LateLintPass<'tcx> for BoolToIntWithIf {
+    fn check_expr(&mut self, ctx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+        if !expr.span.from_expansion() {
+            check_if_else(ctx, expr);
+        }
+    }
+}
+
+fn check_if_else<'tcx>(ctx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+    if_chain!(
+        if let ExprKind::If(check, then, Some(else_)) = expr.kind;
+
+        if let Some(then_lit) = int_literal(then);
+        if let Some(else_lit) = int_literal(else_);
+
+        if check_int_literal_equals_val(then_lit, 1);
+        if check_int_literal_equals_val(else_lit, 0);
+
+        then {
+            let ty = ctx.typeck_results().expr_ty(then_lit); // then and else must be of same type
+
+            let need_parens = should_have_parentheses(check);
+            let snippet = snippet_block(ctx, check.span, "..", None);
+            let snippet_with_braces = format!("{lbrace}{snippet}{rbrace}", lbrace=if need_parens {"("} else {""}, rbrace=if need_parens {")"} else {""});
+
+            span_lint_and_then(ctx, BOOL_TO_INT_WITH_IF, expr.span, "boolean to int conversion using if", |diag| {
+                diag.span_suggestion(
+                    expr.span,
+                    "replace with from",
+                    format!(
+                        "{ty}::from({snippet})"
+                    ),
+                    Applicability::MachineApplicable,
+                );
+
+                diag.note(format!("`{snippet_with_braces} as {ty}` or `{snippet_with_braces}.into()` can also be valid options"));
+            });
+        }
+    );
+}
+
+// If block contains only a int literal expression, return literal expression
+fn int_literal<'tcx>(expr: &'tcx rustc_hir::Expr<'tcx>) -> Option<&'tcx rustc_hir::Expr<'tcx>> {
+    if_chain!(
+        if let ExprKind::Block(block, _) = expr.kind;
+        if let Block {
+            stmts: [],       // Shouldn't lint if statements with side effects
+            expr: Some(expr),
+            hir_id: _,
+            rules: _,
+            span: _,
+            targeted_by_break: _,
```

Comment:
You can also leave these out completely and replace them with `..`. That's usually better, if you only need a few values.

```suggestion
            ..
```

Reply:
Didn't know about that. May be that could be a lint? Seems like it wouldn't be too hard to implement.